File: javascript/src/problems1_100/maximum-subarray.js
class Problem {
  maxSubArray(nums) {
    throw new Error(`Not implemented (${nums}).`);
  }
}

class Default extends Problem {
  maxSubArray(nums) {
    let current = 0;
    let arraySum = nums[0];
    for (const num of nums) {
      current = Math.max(current, 0);
      current += num;
      arraySum = Math.max(arraySum, current);
    }
    return arraySum;
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;

File: python/src/problems1_100/maximum_subarray.py
from abc import ABC, abstractmethod
from typing import override


class Problem(ABC):
    @abstractmethod
    def max_sub_array(self, nums: list[int]) -> int:
        pass


class Default(Problem):
    @override
    def max_sub_array(self, nums: list[int]) -> int:
        current: int = 0
        array_sum: int = nums[0]
        for num in nums:
            current = max(current, 0)
            current += num
            array_sum = max(array_sum, current)
        return array_sum


SOLUTIONS: list[Problem] = [Default()]
